src/mesh/amr_loadbalance.cpp

```diff
@@ -45,6 +45,255 @@
+                        const int i) {
+            fb(t, u, v, k + ks, j + js, i + is) = cb(t, u, v, k, j, i);
+          });
+      // We have to block here w/o buffering so that the write is guaranteed to be
+      // finished before another fine block that is restricted to a sub-region of
+      // this coarse block makes an MPI call and overwrites the coarse buffer.
+      Kokkos::fence();
+    } else {
+#ifdef MPI_PARALLEL
+      PARTHENON_MPI_CHECK(MPI_Recv(var->data.data(), 0, MPI_PARTHENON_REAL, send_rank,
+                                   tag, comm, MPI_STATUS_IGNORE));
+#endif
+    }
+  }
+
+  return test;
+}
+
+#ifdef MPI_PARALLEL
+MPI_Request SendSameToSame(int lid_recv, int dest_rank, Variable<Real> *var,
+                           MeshBlock *pmb, Mesh *pmesh) {
+  MPI_Request req;
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  int tag = CreateAMRMPITag(lid_recv, 0, 0, 0);
+  if (var->IsAllocated()) {
+    // Metadata about this field also needs to be copied from this rank to the
+    // receiving rank (namely the dereference count and the dealloc_count). Not
+    // doing this can cause some subtle differences between runs of the same
+    // problem on different numbers of ranks. Rather than send individual messages,
+    // we just package these in the first two ghost zones of the block. This is ok
+    // because the AMR communication is only required to communicate the interior
+    // zones. The ghosts are communicated to prevent the need to allocate buffers,
+    // but the data in them is not required to be valid. This requires that we have
+    // at least two ghost zones.
+    PARTHENON_REQUIRE(
+        Globals::nghost > 1,
+        "AMR SameToSame communication requires blocks to have at least two ghost zones");
+    auto counter_subview = Kokkos::subview(var->data, std::make_pair(0, 2));
+    auto counter_subview_h = Kokkos::create_mirror_view(HostMemSpace(), counter_subview);
+    counter_subview_h(0) = pmb->pmr->DereferenceCount();
+    counter_subview_h(1) = var->dealloc_count;
+    Kokkos::deep_copy(counter_subview, counter_subview_h);
+
+    PARTHENON_MPI_CHECK(MPI_Isend(var->data.data(), var->data.size(), MPI_PARTHENON_REAL,
+                                  dest_rank, tag, comm, &req));
+  } else {
+    var->com_state[0] = pmb->pmr->DereferenceCount();
+    var->com_state[1] = var->dealloc_count;
+    PARTHENON_MPI_CHECK(
+        MPI_Isend(var->com_state, 2, MPI_INT, dest_rank, tag, comm, &req));
+  }
+  return req;
+}
+
+bool TryRecvSameToSame(int lid_recv, int send_rank, Variable<Real> *var, MeshBlock *pmb,
+                       Mesh *pmesh) {
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  int tag = CreateAMRMPITag(lid_recv, 0, 0, 0);
+
+  int test;
+  MPI_Status status;
+  PARTHENON_MPI_CHECK(MPI_Iprobe(send_rank, tag, comm, &test, &status));
+  if (test) {
+    int size;
+    PARTHENON_MPI_CHECK(MPI_Get_count(&status, MPI_PARTHENON_REAL, &size));
+    if (size > 2) {
```

Comment:
There is no corner case where this could be an issue, is there?
Or would this trigger for 1D sims with `nghost=2`?

Reply:
I 1D think you would have to have `2*nghost + nx <= 2` which I think is impossible.